kori/app/dao/product.py

```diff
@@ -0,0 +1,152 @@
+from kori.app.models import ProductVersion
+from kori.app.models.product import Product
+from kori.app.schemas.product import ProductCreate, ProductSchema, ProductUpdate
+
+settings = Settings()
+
+db_connector = DbConnector(settings.DATABASE_URI)
+
+
+def create(product_data: ProductCreate) -> ProductSchema:
+    session = db_connector.get_session()
+
+    new_product_db = Product(org_id=product_data.org_id, reorder_level=product_data.reorder_level)
+
+    try:
+        session.add(new_product_db)
+        session.flush()
+
+        new_product_version_db = ProductVersion(
+            version_id=1,
+            product_id=new_product_db.id,
+            name=product_data.name,
+            price=product_data.price,
+            measurement_unit=product_data.measurement_unit,
+            valid_from=datetime.now(),
+            valid_to=datetime(9999, 12, 31, 23, 59, 59, 999999),
+        )
+
+        session.add(new_product_version_db)
+        session.commit()
+    except IntegrityError:
+        raise NotFoundException(message="Organisation not found.")
+
+    return ProductSchema(
+        product_id=new_product_db.id,
+        reorder_level=new_product_db.reorder_level,
+        version_id=new_product_version_db.version_id,
+        org_id=new_product_db.org_id,
+        name=new_product_version_db.name,
+        price=new_product_version_db.price,
+        measurement_unit=new_product_version_db.measurement_unit,
+    )
+
+
+def get_product(product_id: UUID, timestamp: Optional[datetime]) -> ProductSchema:
+    session = db_connector.get_session()
+    try:
+        product = session.query(Product).filter(Product.id == product_id).one()
+
+        if product.is_deleted:
+            raise NotFoundException(message="Product is deleted")
+
+        if timestamp:
+            product_version = (
+                session.query(ProductVersion)
+                .filter(ProductVersion.valid_from <= timestamp)
+                .filter(ProductVersion.valid_to >= timestamp)
+                .one()
+            )
+        else:
+            product_version = (
+                session.query(ProductVersion)
+                .filter(ProductVersion.product_id == product_id)
+                .filter(ProductVersion.valid_to == "9999-12-31 23:59:59.999999")
+                .one()
+            )
+
+    except NoResultFound:
+        raise NotFoundException(message="Product Details not found")
+
+    return ProductSchema(
+        product_id=product.id,
+        reorder_level=product.reorder_level,
+        version_id=product_version.version_id,
+        org_id=product.org_id,
+        name=product_version.name,
+        price=product_version.price,
+        measurement_unit=product_version.measurement_unit,
+    )
+
+
+def update(product_id: UUID, product_data: ProductUpdate) -> ProductSchema:
+    session = db_connector.get_session()
+    try:
+        existing_product = session.query(Product).filter(Product.id == product_id).one()
```

Comment:
Note: You can just do `session.query(Product).get(product_id)` instead of `one` when you are filtering by primary keys. Make sure to follow it with a null check

Ref: https://docs.sqlalchemy.org/en/20/orm/queryguide/query.html#sqlalchemy.orm.Query.get